rholang/reference_doc/sends_and_receives/README.md

```diff
@@ -1,17 +1,62 @@
 ## Sends and Receives
 
-### Tuple space
-- Concrete, technical explanation of the tuple space
+### Tuplespace (RSpace)
+Rholang is based on the rhoVM whose computational model is centered around processes sending messages to one another to synchronize when necessary. These operations are coordinated through the tuplespace.
+Although it is unique from the Turing or Von Neuman architectures that many programmers will be familiar with, it is quite understandable. All tuplespace operations can be built up from three building blocks.
+
+![building blocks](buildingBlocks.png)
+
+* A channel on which messages are send and received
+* Data to be sent (includes [primitive types](../primatives/README.md), [data structures](../data_structures/README.md), or even entire processes. )
+* Continuations which spawn new processes when a message is received
 
 ### Sends
-- Content that can be sent
-- Difference between one-tuple and single process in the syntax
-- Persistent sends
-- Send a process (receive a name)
+![Sends](send.png)
+Sending a process associates that process with the channel in the tuplespace. It will wait there forever until it matches with a receive on the same channel.
+
+#### Single Send
+`chan!(data)`
```

Comment:
This looks like a side effect that is not reflected in types. This frightens. No compile time protection. No referential transparency.

Reply:
I'm don't understand this comment.

Reply:
My comment is not specific to this PR. It's more about the overall idea of organizing a system using this non-deterministic basis.

I mean that if you have a referencially transparent function with types, you can predict what's going on based on what you see.
The whole idea of sending data to a channel as a one-way process means that it's non deterministic. The following could happen: (1) current process blocks for unknown period of time in case the other process is not reading the channel; (2) the other process might fail and we won't get the error back; (3) the impact of this statement leaves the scope under consideration. So we cannot understand the program we are looking at without considering the actual wiring with other processes at the level of system. (4) if we drop this statement (or put it under a wrong condition) not much will change from the point of view of the current process. It has no immediate effect on the rest of the computation. (5) If we wrap this statement in a function, we won't know what is happening inside the function without looking into it's definition. And that function will also be non transparent. (6) Testing is much more complex than testing functions. Instead of just checking the result of a function we have to create a mock-object to collect intermediate data emission.

Reply:
Having it non-deterministic is merely a reflection of how the blockchain (and the world it lives in) works. It's non-deterministic by its very nature. You can't tell which parallel statement is going to fire/finish when.

Your questions are more about the rholang language's design and should really be directed do @leithaus.